### DISCUSSION TITLE
ssh: remote control, sessions

### DISCUSSION THREAD

**Discussion by basbebe** [kovidgoyal/kitty#3321] [Q&A]
Hi!

I'm using kitty on a Mac  and I'm interested in how you use remote control over SSH.
The documentation on this seems to be rather sparse so maybe someone using it has some tips / likes to share how he's using it.

Do you use session files to open several ssh windows in one tab?
Or custom shortcuts?

How does remote control over SSH work?
I assume the remote server also has to have kitty installed?
And then I can spawn new windows with hotkeys which will also already be on the remote?
What would I have to set as `listen-on` and `allow-remote` options on the local and remote machine?

I'm using abduco to create persistent sessions on the remote.

One of the servers I SSH to is a FreeBSD machine. Since kitty isn't yet on the FreeBSD ports tree and manually building it wasn't succesfull for me yet, I'm therefore also interested in other FreeBSD users expereience.

**Comment by kovidgoyal** [comment]
On Mon, Feb 15, 2021 at 02:18:45AM -0800, basbebe wrote:
> How does remote control over SSH work?

It works just like remote control locally. Any command run inside the
kitty window can remote control kitty. Or alternately, use a socket and
forward it over SSH.

> I assume the remote server also has to have kitty installed?

No, the remote control protocol is well defined and you can write
your own clients to speak it. https://sw.kovidgoyal.net/kitty/rc_protocol.html

> And then I can spawn new windows with hotkeys which will also already be on the remote?

kitty @ launch ssh hostname

will open a new window connected to the remote server.

> What would I have to set as `listen-on` and `allow-remote` options on the local and remote machine?

there is no allow-remote on the remote end. You just need to run kitty
locally as `kitty -o allow_remote=y`

If you want to connect over a socket, then you use --listen-on and
forward that socket over SSH.

**Reply by basbebe** [reply]
thanks for the quick reply!

On my local machine, I have `KITTY_LISTEN_ON=unix:/tmp/kitty_listen`.

This is the `/usr/share/misc/termcap` on the remote FreeBSD machine:
```
xterm-kitty|KovId's TTY:\
	:tc=xterm-256color:tc=kitty+common:

kitty+common|KovId's TTY common properties:\
	:am:hs:km:mi:ms:xn:\
	:co#80:it#8:li#24:\
	:AL=\E[%dL:DC=\E[%dP:DL=\E[%dM:DO=\E[%dB:IC=\E[%d@:K1=:K3=:\
	:K4=:K5=:LE=\E[%dD:RI=\E[%dC:SF=\E[%dS:SR=\E[%dT:UP=\E[%dA:\
	:ae=\E(B:al=\E[L:as=\E(0:bl=^G:bt=\E[Z:cd=\E[J:ce=\E[K:\
	:cl=\E[H\E[2J:cm=\E[%i%d;%dH:cr=^M:cs=\E[%i%d;%dr:\
	:ct=\E[3g:dc=\E[P:dl=\E[M:do=^J:ds=\E]2;\007:ec=\E[%dX:\
	:ei=\E[4l:fs=^G:ho=\E[H:im=\E[4h:k1=\EOP:k2=\EOQ:k3=\EOR:\
	:k4=\EOS:k5=\E[15~:k6=\E[17~:k7=\E[18~:k8=\E[19~:\
	:k9=\E[20~:kD=\E[3~:kI=\E[2~:kN=\E[6~:kP=\E[5~:kb=\177:\
	:kd=\EOB:ke=\E[?1l:kh=\EOH:kl=\EOD:kr=\EOC:ks=\E[?1h:\
	:ku=\EOA:le=^H:md=\E[1m:me=\E[0m:mh=\E[2m:mr=\E[7m:nd=\E[C:\
	:rc=\E8:..rp=%p1%c\E[%p2%{1}%-%db:\
	:..sa=%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m:\
	:sc=\E7:se=\E[27m:sf=^J:so=\E[7m:sr=\EM:st=\EH:ta=^I:\
	:te=\E[?1049l:ti=\E[?1049h:ts=\E]2;:ue=\E[24m:up=\E[A:\
	:us=\E[4m:vb=\E[?5h\E[?5l:ve=\E[?12l\E[?25h:vi=\E[?25l:\
	:vs=\E[?12;25h:
```

When I do `kitty @ launch` on the remote:
```sh
$ kitty @ launch
sh: kitty: not found
```

Where would I have to forward the socket to to make this work?

Sorry if these questions seem dumb but I'm not sure I understand how this is supposed to work…

**Reply by AprilArcus** [reply]
same question here, not sure how to interpret these instructions

**Reply by groves** [reply]
If you start kitty with `allow_remote_control=yes`, it listens for the terminal escape codes that make up remote control(rc) protocol in the terminal output stream. That's what's described on https://sw.kovidgoyal.net/kitty/rc_protocol/. If anything emits `0x1bP@kitty-cmd` in a kitty with rc enabled, kitty perks up and starts looking for JSON for a command to execute following that.

There's an example emitting those codes using echo a little further down the rc protocol page. It's using a socket, but you could do the same basic thing without a socket like this:

`echo -en '\eP@kitty-cmd{"cmd":"launch","version":[0,14,2],"no_response":true}\e\\'`

`\e` is the echo equivalent of `0x1b` above, so you can see how it's making that same terminal code. 

When you run `kitty @ launch`, it's doing the same thing the echo command does above: emits terminal codes using the rc protocol. It's more convenient than using the echo command as you don't need to remember a bunch of esoteric terminal codes. The drawback is that you need to have kitty installed wherever you're running.

That looks like the issue you're having above, that kitty isn't installed on the server you're SSH'd into. Having a script that uses echo can be more convenient there. You can do anything possible in the rc protocol by changing the JSON command you send via echo.

If you also specify [--listen-on](https://sw.kovidgoyal.net/kitty/invocation/#cmdoption-kitty-listen-on), that changes how kitty listens for rc calls. Instead of watching the terminal output stream, kitty watches that socket. The terminal escape codes are identical, they just need to be sent to the socket instead of printed directly to the terminal.

That's what's happening in this example from the protocol docs:

`echo -en '\eP@kitty-cmd{"cmd":"ls","version":[0,14,2]}\e\\' | socat - unix:/tmp/test`

Rather than echoing directly to standard out, it's being piped into `socat` to get the output to the socket.

The socket option makes it where you must have access to the socket to make an rc command. If you have `allow_remote_control=yes` then any remote host you're SSH'd into can spit out terminal escape codes and drive the kitty on your host computer. Since you can run arbitrary commands with launch, it's pretty straightfoward to go from remote control commands to remote code execution(RCE).

If you get at all nervous about the hosts you're SSHing into being able to run code on your local computer, the socket gives you a middle option to still keep rc on. You can have kitty listen on a socket and only forward the socket to SSH hosts you trust to run rc commands, or don't forward it at all and only allow rc locally.
